test/framework/exec/docker.go

```diff
@@ -0,0 +1,398 @@
+	docker, err := getDockerClient()
+	if err != nil {
+		return err
+	}
+
+	reader, err := docker.ImageSave(ctx, []string{image})
+	if err != nil {
+		return err
+	}
+	defer reader.Close()
+
+	tar, err := os.Create(dest)
+	if err != nil {
+		return err
+	}
+	defer tar.Close()
+
+	_, err = io.Copy(tar, reader)
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+// PullContainerImage triggers the Docker engine to pull an image if it is not
+// already present.
+func PullContainerImage(ctx context.Context, image string) error {
+	docker, err := getDockerClient()
+	if err != nil {
+		return err
+	}
+
+	pullResp, err := docker.ImagePull(ctx, image, types.ImagePullOptions{})
+	if err != nil {
+		return err
+	}
+	defer pullResp.Close()
+
+	// Clients must read the ImagePull response to EOF to complete the pull
+	// operation or errors can occur.
+	_, err = io.ReadAll(pullResp)
+	if err != nil {
+		return nil
+	}
+
+	return nil
+}
+
+// GetLoadBalancerHostPort looks up the host port bound for the load balanced API.
+func GetLoadBalancerHostPort(ctx context.Context, name string) (string, error) {
+	docker, err := getDockerClient()
+	if err != nil {
+		return "", err
+	}
+
+	// Get details about the container
+	containerInfo, err := docker.ContainerInspect(ctx, name)
+	if err != nil {
+		return "", err
+	}
+
+	// Loop through the container port bindings and return the first HostPort
+	for port, bindings := range containerInfo.NetworkSettings.Ports {
+		if port == "6443/tcp" {
+			for _, binding := range bindings {
+				return binding.HostPort, nil
+			}
+		}
+	}
+
+	return "", fmt.Errorf("no host port found for load balancer %q", name)
+}
+
+// ContainerExecToFile executes a command in a running container and writes any output to the provided fileOnHost.
+func ContainerExecToFile(ctx context.Context, containerName string, fileOnHost *os.File, command string, args ...string) error {
+	docker, err := getDockerClient()
+	if err != nil {
+		return err
+	}
+
+	execConfig := types.ExecConfig{
+		Privileged:   true,
+		Cmd:          append([]string{command}, args...),
+		AttachStdout: true,
+		AttachStderr: true,
+	}
+
+	response, err := docker.ContainerExecCreate(ctx, containerName, execConfig)
+	if err != nil {
+		return err
+	}
+
+	execID := response.ID
+	if execID == "" {
+		return fmt.Errorf("exec ID empty")
+	}
+
+	resp, err := docker.ContainerExecAttach(ctx, execID, types.ExecStartCheck{})
+	if err != nil {
+		return err
+	}
+	defer resp.Close()
+
+	// Read out any output from the call
+	outputErrors := make(chan error)
+	go func() {
+		// Send the output to the host file
+		_, err = stdcopy.StdCopy(fileOnHost, fileOnHost, resp.Reader)
+		outputErrors <- err
+	}()
+
+	select {
+	case err := <-outputErrors:
+		if err != nil {
+			return err
+		}
+
+	case <-ctx.Done():
+		return err
+	}
+
+	return nil
+}
+
+// ContainerRunConfig holds the configuration settings for running a container.
+type ContainerRunConfig struct {
```

Comment:
Our pattern in the test framework is usually to call this `RunContainerInput`, but I'm not sure if this is then also preferable here